src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.Keyed.cs

```diff
@@ -0,0 +1,416 @@
+        /// to the <paramref name="collection"/> if the service type hasn't already been registered.
+        /// </summary>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="service">The type of the service to register.</param>
+        /// <param name="serviceKey">The service key.</param>
+        public static void TryAddKeyedScoped(
+            this IServiceCollection collection,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type service,
+            object? serviceKey)
+        {
+            ThrowHelper.ThrowIfNull(collection);
+            ThrowHelper.ThrowIfNull(service);
+
+            var descriptor = ServiceDescriptor.KeyedScoped(service, serviceKey, service);
+            ServiceCollectionDescriptorExtensions.TryAdd(collection, descriptor);
+        }
+
+        /// <summary>
+        /// Adds the specified <paramref name="service"/> as a <see cref="ServiceLifetime.Scoped"/> service
+        /// with the <paramref name="implementationType"/> implementation
+        /// to the <paramref name="collection"/> if the service type hasn't already been registered.
+        /// </summary>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="service">The type of the service to register.</param>
+        /// <param name="serviceKey">The service key.</param>
+        /// <param name="implementationType">The implementation type of the service.</param>
+        public static void TryAddKeyedScoped(
+            this IServiceCollection collection,
+            Type service,
+            object? serviceKey,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type implementationType)
+        {
+            ThrowHelper.ThrowIfNull(collection);
+            ThrowHelper.ThrowIfNull(service);
+            ThrowHelper.ThrowIfNull(implementationType);
+
+            var descriptor = ServiceDescriptor.KeyedScoped(service, serviceKey, implementationType);
+            ServiceCollectionDescriptorExtensions.TryAdd(collection, descriptor);
+        }
+
+        /// <summary>
+        /// Adds the specified <paramref name="service"/> as a <see cref="ServiceLifetime.Scoped"/> service
+        /// using the factory specified in <paramref name="implementationFactory"/>
+        /// to the <paramref name="collection"/> if the service type hasn't already been registered.
+        /// </summary>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="service">The type of the service to register.</param>
+        /// <param name="serviceKey">The service key.</param>
+        /// <param name="implementationFactory">The factory that creates the service.</param>
+        public static void TryAddKeyedScoped(
+            this IServiceCollection collection,
+            Type service,
+            object serviceKey,
```

Comment:
Should this be `object? serviceKey` for pass-through support (like elsewhere in this PR)?